subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy

```diff
@@ -752,6 +798,48 @@ class ArtifactTransformValuesInjectionIntegrationTest extends AbstractDependency
         ann << [Input, InputFile, InputDirectory, OutputFile, OutputFiles, OutputDirectory, OutputDirectories, Destroys, LocalState, OptionValues, Console, Internal]
     }
 
+    @ValidationTestFor(
+        ValidationProblemId.INCOMPATIBLE_ANNOTATIONS
+    )
+    @Unroll
+    def "transform action type cannot use annotation @Untracked for @#propertyType.simpleName"() {
+        settingsFile << """
+            include 'a', 'b'
+        """
+        setupBuildWithColorTransform()
+        buildFile << """
+            project(':a') {
+                dependencies {
+                    implementation project(':b')
+                }
+            }
+
+            abstract class MakeGreen implements TransformAction<TransformParameters.None> {
+                @Untracked
+                @${propertyType.simpleName}
+                abstract Provider<FileSystemLocation> getInputArtifact()
+
+                void transform(TransformOutputs outputs) {
+                    throw new RuntimeException()
+                }
+            }
+"""
+
+        when:
+        fails(":a:resolve")
+
+        then:
+        failure.assertHasDescription('A problem occurred evaluating root project')
+        failure.assertHasCause('A problem was found with the configuration of MakeGreen.')
+        assertPropertyValidationErrors(inputArtifact: incompatibleAnnotations {
+            annotatedWith(Untracked.simpleName)
+            incompatibleWith(propertyType.simpleName)
```

Comment:
Why not pass the class here instead? We are bending over backwards to have this fluent API for testing, I don't see why we should be passing around strings instead of typed objects.

Reply:
Let's discuss this separately. There are various problems around the fixtures for validation messages. I am not sure the above is one, though.